src/diffusers/models/molecule_gnn.py

```diff
@@ -0,0 +1,640 @@
+        super(GINEConv, self).__init__(aggr="add", **kwargs)
+        self.nn = nn
+        self.initial_eps = eps
+
+        if isinstance(activation, str):
+            self.activation = getattr(F, activation)
+        else:
+            self.activation = None
+
+        if train_eps:
+            self.eps = torch.nn.Parameter(torch.Tensor([eps]))
+        else:
+            self.register_buffer("eps", torch.Tensor([eps]))
+
+    def forward(
+        self, x: Union[Tensor, OptPairTensor], edge_index: Adj, edge_attr: OptTensor = None, size: Size = None
+    ) -> torch.Tensor:
+        """"""
+        if isinstance(x, torch.Tensor):
+            x: OptPairTensor = (x, x)
+
+        # Node and edge feature dimensionalites need to match.
+        if isinstance(edge_index, torch.Tensor):
+            assert edge_attr is not None
+            assert x[0].size(-1) == edge_attr.size(-1)
+        elif isinstance(edge_index, SparseTensor):
+            assert x[0].size(-1) == edge_index.size(-1)
+
+        # propagate_type: (x: OptPairTensor, edge_attr: OptTensor)
+        out = self.propagate(edge_index, x=x, edge_attr=edge_attr, size=size)
+
+        x_r = x[1]
+        if x_r is not None:
+            out += (1 + self.eps) * x_r
+
+        return self.nn(out)
+
+    def message(self, x_j: torch.Tensor, edge_attr: torch.Tensor) -> torch.Tensor:
+        if self.activation:
+            return self.activation(x_j + edge_attr)
+        else:
+            return x_j + edge_attr
+
+    def __repr__(self):
+        return "{}(nn={})".format(self.__class__.__name__, self.nn)
+
+
+class GINEncoder(torch.nn.Module):
+    def __init__(self, hidden_dim, num_convs=3, activation="relu", short_cut=True, concat_hidden=False):
+        super().__init__()
+
+        self.hidden_dim = hidden_dim
+        self.num_convs = num_convs
+        self.short_cut = short_cut
+        self.concat_hidden = concat_hidden
+        self.node_emb = nn.Embedding(100, hidden_dim)
+
+        if isinstance(activation, str):
+            self.activation = getattr(F, activation)
+        else:
+            self.activation = None
+
+        self.convs = nn.ModuleList()
+        for i in range(self.num_convs):
+            self.convs.append(
+                GINEConv(
+                    MultiLayerPerceptron(hidden_dim, [hidden_dim, hidden_dim], activation=activation),
+                    activation=activation,
+                )
+            )
+
+    def forward(self, z, edge_index, edge_attr):
+        """
+        Input:
+            data: (torch_geometric.data.Data): batched graph node_attr: node feature tensor with shape (num_node,
+            hidden) edge_attr: edge feature tensor with shape (num_edge, hidden)
+        Output:
+            node_attr graph feature
+        """
+
+        node_attr = self.node_emb(z)  # (num_node, hidden)
+
+        hiddens = []
+        conv_input = node_attr  # (num_node, hidden)
+
```

Comment:
What is node_attr for?

Reply:
Do you mean with respect to the model / application? Or just in this code? Both of them I have only intermediate understanding of.

Reply:
No, I just meant that you do 
```
node_attr = self.node_emb(z)
conv_attr = node_attr
```
why not directly
```
conv_input = self.node_emb(z)
```
?